src/toast/_libtoast/template_offset.cpp

```diff
@@ -182,52 +190,63 @@ void init_template_offset(py::module & m) {
-                            for (
-                                int64_t isamp = dev_intervals[iview].first;
-                                isamp <= dev_intervals[iview].last;
-                                isamp++
-                            ) {
-                                int64_t d = data_index * n_samp + isamp;
-                                int64_t amp = amp_offset + amp_view_off[iview] +
-                                              (int64_t)(
-                                    (isamp - dev_intervals[iview].first) / step_length
-                                              );
-                                contrib = 0.0;
+                        for (int64_t isamp = 0; isamp < max_interval_size;
+                             isamp += step_length) {
+                            // Adjust for the actual start of the interval
+                            int64_t adjusted_isamp = isamp + dev_intervals[iview].first;
+
+                            // Check if the value is out of range for the current
+                            // interval
+                            if (adjusted_isamp > dev_intervals[iview].last) {
+                                continue;
+                            }
+
+                            // Insure we do not go out of the current interval
+                            int64_t max_step_length =
+                                std::min(step_length,
+                                         dev_intervals[iview].last - adjusted_isamp + 1);
+
+                            // Reduce on a chunk of `step_length` samples.
+                            double contrib = 0.0;
+                            # pragma omp parallel for reduction(+ : contrib)
+                            for (int64_t i = 0; i < max_step_length; i++) {
+                                int64_t d = data_index * n_samp + adjusted_isamp + i;
                                 if (use_flags) {
-                                    int64_t f = flag_index * n_samp + isamp;
+                                    int64_t f = flag_index * n_samp + adjusted_isamp + i;
                                     uint8_t check = dev_det_flags[f] & flag_mask;
                                     if (check == 0) {
-                                        contrib = dev_det_data[d];
+                                        contrib += dev_det_data[d];
                                     }
                                 } else {
-                                    contrib = dev_det_data[d];
+                                    contrib += dev_det_data[d];
                                 }
-                                # pragma omp atomic update
-                                dev_amplitudes[amp] += contrib;
                             }
+
+                            int64_t amp = amp_offset + amp_view_off[iview] +
+                                          (int64_t)(isamp / step_length);
+                            dev_amplitudes[amp] += contrib;
```

Comment:
The threaded reduction to the team-local `contrib` variable looks fine, and for this particular template, the contribution to an amplitude is localized to just this range of samples from one detector.  So no need for atomic update.  Great!